[GitHub Repository Search] Add dropdown for filtering by user/org

jberglinds opened the pull request:
## Description
Adds a dropdown to the search field for filtering the search by user/orgs, similarly to how the native "Search Repositories" extension works. Also, includes forks when searching for repositories like the native extension.

* "All Repositories" -> The same as before, searching all of Github (Default selection)
* "My Repositories" -> Searches for personal repos and repos of organizations that you are a member of
* "{Github username}" -> Searches only for personal repos
* "{Organization username}" -> Searches only for repos of the chosen org

## Screencast
<img width="1000" alt="github-repository-search 2022-07-14 at 16 05 04" src="https://user-images.githubusercontent.com/6368728/179001444-ac2d4032-f6f0-44a7-948f-56863b5f514b.png">
<img width="1000" alt="github-repository-search 2022-07-14 at 16 05 56" src="https://user-images.githubusercontent.com/6368728/179001632-40d9458e-db11-43a3-9be2-8e70ee5b2a02.png">
<img width="1000" alt="github-repository-search 2022-07-14 at 16 03 31" src="https://user-images.githubusercontent.com/6368728/179001121-ddcf6217-de42-4091-8d5b-f91f59b5b153.png">


<!-- If you add a new extension or command, include a screencast (or screenshot for straightforward changes). A good screencast will make the review much faster - especially if your extension requires registration in other services.  -->

## Checklist

- [ ] I read the [extension guidelines](https://developers.raycast.com/basics/prepare-an-extension-for-store)
- [ ] I read the [documentation about publishing](https://developers.raycast.com/basics/publish-an-extension)
- [x] I ran `npm run build` and [tested this distribution build in Raycast](https://developers.raycast.com/basics/prepare-an-extension-for-store#metadata-and-configuration)
- [ ] I checked that files in the `assets` folder are used by the extension itself
- [ ] I checked that assets used by the `README` are placed outside of the `metadata` folder

jberglinds commented on extensions/github-repository-search/src/index.tsx:
```diff
@@ -95,7 +153,7 @@ function Actions(props: { repository: Repository; onVisit: (repository: Reposito
           title="Open Pull Requests"
           url={`${props.repository.url}/pulls`}
-          shortcut={{ modifiers: ["cmd"], key: "p" }}
+          shortcut={{ modifiers: ["cmd", "shift"], key: "p" }}
```
cmd+p is the default hotkey to trigger the dropdown so this had to change

thomaspaulmann replied:
Makes sense, thanks!

jberglinds commented on extensions/github-repository-search/src/index.tsx:
```diff
@@ -121,7 +179,7 @@ function Actions(props: { repository: Repository; onVisit: (repository: Reposito
             title="Open Projects"
             url={`${props.repository.url}/projects`}
-            shortcut={{ modifiers: ["cmd", "shift"], key: "p" }}
+            shortcut={{ modifiers: ["cmd", "shift", "opt"], key: "p" }}
```
this is quite a lot, open to suggestions

thomaspaulmann replied:
I guess it's fine.

jberglinds commented on extensions/github-repository-search/src/index.tsx:
```diff
@@ -11,55 +11,113 @@ import {
+            (r) =>
+              r.nameWithOwner.includes(searchText ?? "") &&
+              r.nameWithOwner.match(
+                // Converting query filter string to regexp:
+                `${searchFilter?.replaceAll(/org:|user:/g, "").replaceAll(" ", "|")}\/.*`
```
not proud of this but the alternative is quite some extra state to manage

DonDebonair commented:
This looks great! How can we help test this?

jberglinds commented:
As this week is "launch week" at raycast I'm guessing that Thomas is really busy. 

If you want to test it out before it gets merged you can clone the changes and run `npm install && npm build` and it should install as a separate extension, @DonDebonair

DonDebonair commented:
I tested it (need to run `npm run dev` to get it to work btw) and it works great! The added option to search only your own repos or even within a specific org, together with caching visited repos, is exactly what I need!

Would be nice if this could be merged and published soon :)

thomaspaulmann commented:
Sorry that this drags so long. Been busy with Launch Week. Gonna pick it up next week.

jberglinds commented:
Hey @thomaspaulmann, any chance you could take a look at this soon-ish?

thomaspaulmann commented on extensions/github-repository-search/tsconfig.json:
```diff
@@ -3,7 +3,7 @@
   "include": ["src/**/*"],
   "compilerOptions": {
-    "lib": ["es2020"],
+    "lib": ["es2021"],
```
Why did you change this?

jberglinds replied:
```
src/index.tsx:74:34 - error TS2550: Property 'replaceAll' does not exist on type 'string'. Do you need to change your target library? Try changing the 'lib' compiler option to 'es2021' or later.
```

thomaspaulmann commented on extensions/github-repository-search/src/useRepositories.ts:
```diff
@@ -53,7 +53,9 @@ export function useRepositories(searchText: string | undefined) {
       try {
-        const { search } = await octokit.graphql<SearchRepositoriesResponse>(SEARCH_REPOSITORIES_QUERY, { searchText });
+        const { search } = await octokit.graphql<SearchRepositoriesResponse>(SEARCH_REPOSITORIES_QUERY, {
+          searchQuery: `${searchFilter} fork:true ${searchText}`,
```
Should it be a preference if you want to include forks in the search? It's a bit confusing that you have `searchText` and `searchFilter`. Wouldn't it be easier to inject the entire `searchQuery` from outside the hook?

jberglinds replied:
Sure, can do it

thomaspaulmann commented on extensions/github-repository-search/src/index.tsx:
```diff
@@ -11,55 +11,113 @@ import {
+  userData,
+  onFilterChange,
+}: {
+  userData: UserDataResponse["viewer"] | undefined;
```
Isn't that the `viewer`?
```suggestion
  viewer?: UserDataResponse["viewer"];
```

jberglinds replied:
Sure it is.
I renamed it because I didn't think 'viewer' was a great name anymore when you detach it from the context of the request, but I guess you still have it in the type information

thomaspaulmann commented on extensions/github-repository-search/src/index.tsx:
```diff
@@ -11,55 +11,113 @@ import {
+  userData: UserDataResponse["viewer"] | undefined;
+  onFilterChange: (filter: string) => void;
+}) {
+  return (
+    <List.Dropdown tooltip="Filter Repositories" storeValue={true} onChange={onFilterChange}>
+      <List.Dropdown.Item title={"All Repositories"} value={""} />
+      {userData && (
+        <List.Dropdown.Item
+          title={"My Repositories"}
+          value={`user:${userData.login} ${userData.organizations.nodes.map((org) => `org:${org.login}`).join(" ")}`}
+        />
+      )}
+      {userData && (
+        <List.Dropdown.Section>
+          <List.Dropdown.Item title={userData.login} value={`user:${userData.login}`} />
+          {userData.organizations.nodes.map((org) => (
+            <List.Dropdown.Item key={org.login} title={org.login} value={`org:${org.login}`} />
+          ))}
+        </List.Dropdown.Section>
+      )}
+    </List.Dropdown>
+  );
+}
```
How about not returning a dropdown if we don't have any user data? A dropdown with just a single item doesn't make much sense.
```suggestion
  return userData ? (
    <List.Dropdown tooltip="Filter Repositories" storeValue={true} onChange={onFilterChange}>
      <List.Dropdown.Item title="All Repositories" value="" />
        <List.Dropdown.Item
          title="My Repositories"
          value={`user:${userData.login} ${userData.organizations.nodes.map((org) => `org:${org.login}`).join(" ")}`}
        />
        <List.Dropdown.Section>
          <List.Dropdown.Item title={userData.login} value={`user:${userData.login}`} />
          {userData.organizations.nodes.map((org) => (
            <List.Dropdown.Item key={org.login} title={org.login} value={`org:${org.login}`} />
          ))}
        </List.Dropdown.Section>
    </List.Dropdown> : null
  );
}
```

thomaspaulmann commented on extensions/github-repository-search/src/index.tsx:
```diff
@@ -11,55 +11,113 @@ import {
-          ?.filter((r) => r.nameWithOwner.includes(searchText ?? ""))
+          ?.filter(
+            (r) =>
+              r.nameWithOwner.includes(searchText ?? "") &&
+              r.nameWithOwner.match(
+                // Converting query filter string to regexp:
+                `${searchFilter?.replaceAll(/org:|user:/g, "").replaceAll(" ", "|")}/.*`
+              )
```
Shouldn't that be a disjunction?
```suggestion
              r.nameWithOwner.includes(searchText ?? "") ||
              r.nameWithOwner.match(
                // Converting query filter string to regexp:
                `${searchFilter?.replaceAll(/org:|user:/g, "").replaceAll(" ", "|")}/.*`
              )
```

jberglinds replied:
Why?
The second filter just targets the owner part of the repository id. Even if you get a hit with the first filter on something I think it should still filter the results by whatever you have selected in the dropdown

thomaspaulmann reviewed (changes requested):
I left a few comments in the PR. My main concern is the new required scopes. People that don't have a PAT with the user and org scope will constantly get failure toasts. Given that the extension is used by a lot of people, we should find a way to work around this.

<img width="1000" alt="github-repository-search 2022-07-29 at 10 37 46" src="https://user-images.githubusercontent.com/12066405/181731822-3ce03798-0148-4716-be75-66a8ce85be5b.png">

A few ideas:
- Somehow check the scope and do the right thing. If user/org scope is available, add your changes, if not, then stick with the previous behavior.
- Put your changes behind a new preference that people have to opt-in. I don't like preferences for such things.

We have this problem with other extensions and need some elegant way to solve it. I'm open for suggestions here.

thomaspaulmann commented:
@jberglinds left a few comments. Sorry that it took me longer than expected.

jberglinds commented:
I believed I've addressed most of your comments. 
Regarding the error messages I'm not sure what to do about it. I think the error message it pretty clear about whats wrong.

thomaspaulmann commented:
It isn't about that the error message isn't clear. It's about that existing users see this message from one day to another. This causes confusion.

I have an idea to be smarter:
- Show a one-time view to explain the new features and mention that you need a token with the expected scopes
- Build the API requests in a more robust way
  - Essentially, if the user doesn't have user scope, don't search private repos. And if the user's scope doesn't have access to orgs, don't search them.
  - The dropdown would then be optional.

Benefit is that existing users can keep using the extension as is and don't need to change the scope of the token. Happy to help implementing this. WDYT?

jberglinds commented:
I've had absolutely no time lately to look at this. Please help out if you can find the time 🙏

thomaspaulmann commented:
I started to work on the changes in [this branch](https://github.com/raycast/extensions/tree/pr/2266). For now I upgraded to the latest APIs. Next, I'll convert the history to use our utils and then I'll look into the graceful handling of the tokens.

thomaspaulmann commented:
@jberglinds I found the time to upgrade the extension. I created a new PR for it, it was easier after all the additional changes. I'm closing this PR for #2901. Please give it a try and let me know if there is anything else that you would need.
